File: C++/Variables.cpp
#include <iostream>
using namespace std;

int main() {
    // =====================
    // Variables in C++
    // =====================

    // ======= 1. تعريف المتغيرات =======
    int age = 25;              // متغير من نوع integer لتخزين الأعداد الصحيحة
    double pi = 3.14159;       // متغير من نوع double لتخزين الأعداد العشرية
    float temperature = 36.6;  // متغير من نوع float (أقل دقة من double)
    char grade = 'A';           // متغير من نوع char لتخزين حرف واحد
    string name = "أسامة";     // متغير من نوع string لتخزين نصوص
    bool isStudent = true;      // متغير من نوع bool لتخزين القيمتين true أو false

    // ======= 2. طباعة المتغيرات =======
    cout << "الاسم: " << name << endl;
    cout << "العمر: " << age << endl;
    cout << "درجة الطالب: " << grade << endl;
    cout << "درجة الحرارة: " << temperature << endl;
    cout << "قيمة PI: " << pi << endl;
    cout << "هل هو طالب؟ " << isStudent << endl;

    // ======= 3. تغيير قيمة المتغيرات =======
    age = 26;  // تحديث العمر
    pi = 3.14; // تحديث قيمة PI
    cout << "\nبعد التحديث:" << endl;
    cout << "العمر: " << age << endl;
    cout << "قيمة PI: " << pi << endl;

    // ======= 4. أفضل الممارسات =======
    // - استخدم أسماء واضحة للمتغيرات تصف محتواها
    // - اختر النوع المناسب للمتغير (int, double, float, char, string, bool)
    // - لا تترك متغيرات بدون تهيئة إذا كنت ستستخدمها مباشرة
    // - ضع تعليقات لتوضيح سبب استخدام المتغير أو نوعه إذا كان غير واضح

    return 0;
}

File: dart/Variables.dart
// =====================
// Variables in Dart
// =====================

// ignore_for_file: avoid_print

void main() {
  // ======== المتغيرات العادية ========
  var name = 'Osama'; // النوع يُستنتج تلقائياً
  print('Name: $name');

  String country = 'Saudi Arabia'; // تحديد النوع صراحة
  print('Country: $country');

  int age = 22;
  print('Age: $age');

  double pi = 3.14;
  print('Pi: $pi');

  bool isStudent = true;
  print('Is student? $isStudent');

  // ======== القوائم (List) ========
  List<int> numbers = [1, 2, 3, 4];
  print('Numbers: $numbers');

  // ======== القواميس (Map) ========
  Map<String, int> scores = {'Ali': 90, 'Sara': 85};
  print('Scores: $scores');

  // ======== الثوابت (final & const) ========
  final today = DateTime.now(); // يُحدد وقت التشغيل
  print('Today: $today');

  const gravity = 9.8; // ثابت وقت الترجمة
  print('Gravity: $gravity');

  // ======== المتغيرات Nullable ========
  String? nickname; // ممكن تكون null
  print('Nickname: $nickname');

  nickname = 'Sam';
  print('Updated nickname: $nickname');

  // ======== الكلمة المفتاحية late ========
  late String description;
  description = 'This variable is initialized later.';
  print(description);

  // ======== الكلمة المفتاحية dynamic ========
  dynamic anything = 'Hello';
  print('Dynamic: $anything');

  anything = 12345; // ممكن يتغير النوع
  print('Dynamic changed: $anything');
}
